fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+    only_relayer(env, state)?;
+
+    if amount > state.swap_fee {
+        // TOFIX(LR) when amount == fee, amount will still be consumed
+        let swap_fee = state.swap_fee;
+        let effective_amount = (amount - swap_fee)?;
+        let to_canonical = deps.api.canonical_address(&to)?;
+        let rtx = send_tokens(
+            &deps.api,
+            &env,
+            &state,
+            &to_canonical,
+            amount,
+            "reverse_swap",
+        )?;
+        config(&mut deps.storage).update(|mut state| {
+            state.supply = (state.supply - amount)?;
+            state.sealed_reverse_swap_id += 1; // TOFIX(LR) after exec should be == rid
+                                               //state.sealed_reverse_swap_id = rid;
+            Ok(state)
+        })?;
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: rtx.messages,
+            log,
+            data: None,
+        };
+        Ok(r)
+    } else {
+        let swap_fee = amount;
+        let effective_amount = Uint128::zero();
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: vec![],
+            log,
+            data: None,
+        };
+        Ok(r)
+    }
+}
+
+/* ***************************************************
+ * *****************    Helpers      *****************
+ * ***************************************************/
+
+fn amount_from_funds(funds: &Vec<Coin>) -> Uint128 {
+    // TODO(LR) return proper error
+    if funds.len() == 1
+    /* && funds[0].denom == "fet" */
+    {
+        // TODO(LR) does cosmwas allows sending multiple funds of the same token
+        funds[0].amount
+    } else {
+        Uint128::zero()
+    }
+}
+
+fn send_tokens<A: Api>(
+    api: &A,
+    env: &Env,
+    state: &State,
+    to_address: &CanonicalAddr,
+    amount: Uint128,
+    action: &str,
+) -> HandleResult {
+    let from_address = &env.contract.address;
+    let from_human = api.human_address(&from_address)?;
+    let to_human = api.human_address(to_address)?;
+    let log = vec![log("action", action), log("to", to_human.as_str())];
+    let coin = Coin {
+        amount,
+        denom: state.denom.clone(),
+    };
+
+    let r = HandleResponse {
+        messages: vec![CosmosMsg::Bank(BankMsg::Send {
+            from_address: from_human,
+            to_address: to_human,
+            amount: vec![coin],
+        })],
```

Comment:
Question is, whether it is actually really ensured, that if the `BankMsg::Send` message fails for whatever reason (e.g. not enough funcds on contract address), then whole contract state change is reverted?

I'm asking, because this is new message dynamically generated by this contract during contract call execution => it is essential to be sure that whole contract call will fail (and state change will be reverted) if this dynamic Send message fails to execute.

Reply:
@pbukva that's a good point, I just assume it will be the case.
I just tested it manually and the state is properly reverted.
but I need to find a reference that confirms that.

just to make sure we are on the same page, I am not executing the `::Send` in the contract, I am just returning the message for the send to be processed by validator and included into tx)

Reply:
I understood that - and that is exactly why barking on this tree - since it sort of feels like execution of this dynamic Send message might not be (potentially) related to execution of the cottract call ...